### DISCUSSION TITLE
Track numbers above 255

### DISCUSSION THREAD

**Discussion by DavidAIsom** [Makuna/DFMiniMp3#104] [Q&A] [answered]
Hi I am struggling to get track numbers above 255 to be returned by the player . I can call them and play them by explicite requesting but can’t seem to get the module to report back via the  uint16_t getCurrentTrack(DfMp3_PlaySource source) command 
or uint16_t getTotalTrackCount(DfMp3_PlaySource source)

Ay ideas ?

**Comment by Makuna** [comment]
(moved to Discussions)

What module do you have?  
Can you read the chip number on it?  
Which chip variant are you defining in your code (Mp3ChipMH2024K16SS or Mp3ChipOriginal)?
Where do you place the songs?  Mp3 folder?  Root?  Numbered Folder?

Some APIs calls have limits to 255 tracks.  This tends to be the "folder" APIs.
This is from the spec...
```
  Supports maximum 3000 audio files in the root directory of the storage device. 
  Supports maximum 99 folders, and each folder can stores 255 audio files.
```
Another chip states a "max of 1000 songs".

**Reply by DavidAIsom** [reply]
I am using  mp3 folder.  No problem playing 3000 tracks it’s the reporting and content that’s the question

**Comment by Makuna** [comment]
What Arduino platform are you using?  (still some other questions you haven't answered).
The unreleased next version on Github has fixes for AVR.

**Reply by DavidAIsom** [reply]
The hardware it’s connected to is a mega 2560 is that what you are asking. ?

**Reply by Makuna** [reply]
Yes, AVR is the general name of the platform, Mega2560 is the specific name.
You will want to grab the github master and try it.  Arduino's last AVR update messed with the code generation causing some issues that don't show up on other platforms.

**Comment by DavidAIsom** [comment]
As far as I know I have all the latest updates Sent from my iPadOn 6 Feb 2023, at 22:13, Michael Miller ***@***.***> wrote:﻿
Yes, AVR is the general name of the platform, Mega2560 is the specific name.
You will want to grab the github master and try it.  Arduino's last AVR update messed with the code generation causing some issues that don't show up on other platforms.

**Reply by Makuna** [reply]
You have the latest released version?  The latest Master version (version number increments when it is released, so it will have the same number as the released until then) is only accessible by either using Git and clone it locally or by downloading and copying files to specific locations.  It has not been released and has changes that affect AVR.

**Comment by DavidAIsom** [comment]
Please can you point me to the downloads that I should use ?Sent from my iPadOn 6 Feb 2023, at 23:20, Michael Miller ***@***.***> wrote:﻿
You have the latest released version?  The latest Master version (version number increments when it is released, so it will have the same number as the released until then) is only accessible by either using Git and clone it locally or by downloading and copying files to specific locations.  It has not been released and has changes that affect AVR.

**Reply by Makuna** [reply]
https://github.com/Makuna/DFMiniMp3
You will see a CODE button, which opens a popup menu, with several options, download and then, copy contents into your arduino libraries folder (C:\Users\<user name>\Documents\Arduino\libraries\DFMiniMp3), which is different for Mac, but you will find the current DFMiniMp3 folder which you can replace.

**Reply by DavidAIsom** [reply]
Hi , I have uploaded and installed this version. The number of records still reports up to 255 ok. If there are more then 255 it reports the difference ie 284 reports 28 David

**Comment by DavidAIsom** [comment]
Here’s photos of the module Sent from my iPhoneOn 7 Feb 2023, at 00:54, Michael Miller ***@***.***> wrote:﻿
https://github.com/Makuna/DFMiniMp3
You will see a CODE button, which opens a popup menu, with several options, download and then, copy contents into your arduino libraries folder (C:\Users<user name>\Documents\Arduino\libraries\DFMiniMp3), which is different for Mac, but you will find the current DFMiniMp3 folder which you can replace.

**Reply by Makuna** [reply]
Using Email doesn't attach the photo.

**Comment by DavidAIsom** [comment]
How can I send photo ?Sent from my iPadOn 7 Feb 2023, at 17:04, Michael Miller ***@***.***> wrote:﻿
Using Email doesn't attach the photo.

**Reply by DavidAIsom** [reply]
Sorted here they are … next message

**Comment by DavidAIsom** [comment]
![C8C4E7A9-EA5E-4CE5-A98F-51FA62B1C536](https://user-images.githubusercontent.com/58168488/217323447-be2744dc-e84a-4449-93c0-e82be9017797.jpeg)
![28C7C56A-7349-4FC5-BC2E-D8361B474A83](https://user-images.githubusercontent.com/58168488/217323465-8975e0d2-edc7-4a09-950f-96959cf84a2a.jpeg)

**Comment by Makuna** [comment]
So, have you tried using the Mp3ChipMH2024K16SS in your declaration of your mp3 object?  Notice the chip is a MH2024K.

It will take me a while to try to replicate your scenario.

**Comment by DavidAIsom** [comment]
Want quite sure how to it …looking  at your .h file the device listed is not quite the same.  Please can you give me a clue on how to do this ?

**Comment by DavidAIsom** [comment]
Just wanted to say I really appreciate that you are trying to help with this.David Sent from my iPadOn 7 Feb 2023, at 18:51, Michael Miller ***@***.***> wrote:﻿
So, have you tried using the Mp3ChipMH2024K16SS in your declaration of your mp3 object?  Notice the chip is a MH2024K.
It will take me a while to try to replicate your scenario.

**Comment by Makuna** [comment]
In your sketch, you will define the mp3 object from the examples like 
```
typedef DFMiniMp3<HardwareSerial, Mp3Notify> DfMp3; 
```
this defalts to the original and is the same as
```
typedef DFMiniMp3<HardwareSerial, Mp3Notify, Mp3ChipOriginal> DfMp3; 
```
to use the alternative, then just change that template argument to one I mentioned like
```
typedef DFMiniMp3<HardwareSerial, Mp3Notify, Mp3ChipMH2024K16SS> DfMp3; 
```

I am spending more effort on this issue as I am concerned it is related to the AVR compiler issue (byte to word boundary issues).  ;-)

**Comment by DavidAIsom** [comment]
Implemented as suggested same result reports the excess over 255/256 as beforeDavid Sent from my iPadOn 7 Feb 2023, at 20:03, Michael Miller ***@***.***> wrote:﻿
In your sketch, you will define the mp3 object from the examples like
typedef DFMiniMp3<HardwareSerial, Mp3Notify> DfMp3; 

this defalts to the original and is the same as
typedef DFMiniMp3<HardwareSerial, Mp3Notify, Mp3ChipOriginal> DfMp3; 

to use the alternative, then just change that template argument to one I mentioned like
typedef DFMiniMp3<HardwareSerial, Mp3Notify, Mp3ChipMH2024K16SS> DfMp3; 

I am spending more effort on this issue as I am concerned it is related to the AVR compiler issue (byte to word boundary issues).  ;-)

**Comment by Makuna** [comment, marked as answer]
Finally put 1000 mp3 files into the mp3 directory on the SD card and tried to reproduce.

```
uint16_t count = dfmp3.getTotalTrackCount(DfMp3_PlaySource_Sd);
  Serial.print("files ");
  Serial.println(count);
```
returned 1000 correctly.
```
OUT  7e ff 06 48 00 00 00 fe b3 ef
IN  7e ff 06 48 00 03 e8 fd c8 ef
files 1000
```
The notification handler like...
```
static void OnPlayFinished([[maybe_unused]] DfMp3& mp3, [[maybe_unused]] DfMp3_PlaySources source, uint16_t track)   {
    Serial.print("Play finished for #");
    Serial.println(track);  

    track = dfmp3.getCurrentTrack();
    Serial.print("Current Track #");
    Serial.println(track);  
}
```
also returned track numbers above 255 correctly, and the call to get current track above also returned the next track that was playing correctly.
```
IN  7e ff 06 3d 00 01 70 fe 4d ef
Play finished for #368
OUT  7e ff 06 4c 00 00 00 fe af ef
IN  7e ff 06 4c 00 01 71 fe 3d ef
Current Track #369
```

These extended debug output seen above starting with IN and OUT I just added, so you will have to pick up the very latest.
To enable them, just add this line above the include for the library like...
```
#define DfMiniMp3Debug Serial

#include <DFMiniMp3.h>
```

**Reply by DavidAIsom** [reply]
I looked for avr code here 

https://github.com/arduino/ArduinoCore-avr/releases

tried tto download and install got this message    Error: 13 INTERNAL: Library install failed: library not valid

**Comment by Makuna** [comment]
BTW, my test is with only files in the MP3 folder, named 0000_anything.mp3 as required.  Note starts with four digits zero padded.  Having files in multiple locations is problematic as the API is not good with mixed use.  Like, getCurrentTrack, is that from root, mp3, numbered folder (01,02), or advert folders?  They aren't specific in the documents.

**Comment by DavidAIsom** [comment]
All my files are in mp3  folder and named    nnnn.mp3 Sent from my iPadOn 8 Feb 2023, at 02:06, Michael Miller ***@***.***> wrote:﻿
BTW, my test is with only files in the MP3 folder, named 0000_anything.mp3 as required.  Note starts with four digits zero padded.  Having files in multiple locations is problematic as the API is not good with mixed use.  Like, getCurrentTrack, is that from root, mp3, numbered folder (01,02), or advert folders?  They aren't specific in the documents.

**Comment by DavidAIsom** [comment]
HI,  Thank you for this here is my result

07:58:39.056 -> OUT  7e ff 06 48 00 00 00 fe b3 ef

07:58:39.134 -> IN  7e ff 06 48 00 00 a3 fe 10 ef

07:58:39.134 -> files 163


I actually have 3748 files as you can see high byte not populated


how can i check i have the latest avr for mega 2560 have looked on
github tried several downloads none install


david

**Reply by Makuna** [reply]
Use the "board manager" in Arduino IDE, find "Arduino AVR Boards", the latest is v1.8.6

I would suggest you try just having 1000 files and see what happens.  If that works like mine, then try 2000, then 3000, and narrow it down to find out the count at which the problem happens.

**Comment by DavidAIsom** [comment]
The problem starts at 256 . Sent from my iPhoneOn 8 Feb 2023, at 16:20, Michael Miller ***@***.***> wrote:﻿
Use the "board manager" in Arduino IDE, find "Arduino AVR Boards", the latest is v1.8.6
I would suggest you try just having 1000 files and see what happens.  If that works like mine, then try 2000, then 3000, and narrow it down to find out the count at which the problem happens.

**Comment by DavidAIsom** [comment]
It accepts them for playing but just won’t report any tracks above 255 Sent from my iPhoneOn 8 Feb 2023, at 16:20, Michael Miller ***@***.***> wrote:﻿
Use the "board manager" in Arduino IDE, find "Arduino AVR Boards", the latest is v1.8.6
I would suggest you try just having 1000 files and see what happens.  If that works like mine, then try 2000, then 3000, and narrow it down to find out the count at which the problem happens.

**Reply by Makuna** [reply]
SO, to be clear, you placed ONLY 1000 mp3 files on the SD card and the problem persists?

**Comment by DavidAIsom** [comment]
With 256 it reports 0  with 257 it reports 1 at 510 or reports 255 and so on Sent from my iPhoneOn 8 Feb 2023, at 16:30, Michael Miller ***@***.***> wrote:﻿
SO, to be clear, you placed ONLY 1000 mp3 files on the SD card and the problem persists?

**Reply by DavidAIsom** [reply]
basically it counts perfectly up to 255 then resets to zero  and starts again over and over.  its the upport byte that does not repor.t i have copied 255, 256 257, then various up to 300 400 etc its clear the upper byte is the issue as far as reporting back.  i can play any track number i want ad they play fine. its just that when play finishes its only reports lower byte.

**Comment by DavidAIsom** [comment]
do  you thtink it could be the module itself./ iave 2 and they are both  the same.

**Reply by Makuna** [reply]
Yes, it's the module.  The raw data returned from those extra debug statements is what the module is sending; and its wrong.
I have heard there is a way to flash the modules with updates.  I have never tried it; but maybe that will correct their problem.

**Comment by DavidAIsom** [comment]
I am happy to purchase a different one. It’s just finding a maker who can supply good device . Any ideas ? Sent from my iPadOn 8 Feb 2023, at 17:10, Michael Miller ***@***.***> wrote:﻿
Yes, it's the module.  The raw data returned from those extra debug statements is what the module is sending; and its wrong.
I have heard there is a way to flash the modules with updates.  I have never tried it; but maybe that will correct their problem.

**Comment by DavidAIsom** [comment]
Hi Again,

I took ordered new module from amazon yesterday.  It has just arrvied and on conection ...... it works perfectly !!!

18:23:05.856 -> initializing...
18:23:05.936 -> volume 25
18:23:06.123 -> files 3747
18:23:06.123 -> starting..

I do want you to know  how much you helped. 

 Hope you are OK with having made those changes ? 

David

**Reply by Makuna** [reply]
Great news!  Can you post an image of the chipset?

These modules have been rather erratic in functionality.  I think the first ones I got were basically GEN II, as the original docs were already wrong on about 10% of the API and even the official WIKI on DFRobot has incorrect information from GEN I chips even though their own code demonstrates GENII APIs ;-)

**Comment by DavidAIsom** [comment]
![5A34E71E-B38B-4BA2-9DC4-46F971BCB975](https://user-images.githubusercontent.com/58168488/217899409-471b7947-b738-43d0-b5d8-08c01a6b3042.jpeg)
![100D720F-4830-4013-B162-BC474FEAA6D1](https://user-images.githubusercontent.com/58168488/217899482-29449b87-3f7d-4c1b-ab48-5a395847f6d5.jpeg)
![B5C91337-CD82-41A1-83E4-4D05A7C0147A](https://user-images.githubusercontent.com/58168488/217899490-61c2dc4d-bb35-4634-8c2f-7a93329575ef.jpeg)
![5420FA43-F083-4878-819A-34E1E9693DBB](https://user-images.githubusercontent.com/58168488/217899496-c682e6f1-d995-4c55-89aa-69bba2e70f32.jpeg)
![AC82F0BD-13EB-4ADD-B53C-EEE138E8D2AB](https://user-images.githubusercontent.com/58168488/217899519-32260dda-a8ce-40f4-97ef-37ababe52630.jpeg)
